### PR TITLE
[0.4.x] Vite 3 support

### PR DESCRIPTION
**Pull request by timacdonald**
## Stack Testing

- [x] Vanilla skeleton install
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
- [x] Breeze w/ blade
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
- [x] Breeze w/ vue
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
- [x] Breeze w/ vue w/ ssr
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
  - [x] SSR
- [x] Breeze w/ react
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
- [x] Breeze w/ react w/ ssr
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
  - [x] SSR
- [x] Jetstream w/ livewire
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
- [x] Jetstream w/ inertia
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
- [x] Jetstream w/ inertia w/ ssr
  - [x] Hot reloading (CSS + JS)
  - [x] Blade Refreshing
  - [x] SSR

## Servers

- [x] Artisan Serve
  - [x] Hot reloading
  - [x] Blade Refreshing

- [x] Valet
  - [x] Hot reloading
  - [x] Blade Refreshing

- [x] Sail
  - [x] Hot reloading
  - [x] Blade Refreshing

- [x] Secured sites

## Misc

- [x] Inertia helpers are not externalised in SSR build (temp fix in place)
- [x] Investigate resolveServerUrls.local[0]
- [x] Vite dev server "this is not the page you are looking for 🤚" 
  - [x] browser page
  - [x] console logging
- [x] Investigate why full page refresh is triggered on `storage/framework/views/**` changes, triggering a double refresh.
- [x] It is possible to get the "this is not the page you are looking for" console log when the dev server pinged by the front-end. We should potentially check if the headers are requesting HTML
- [x] Update `globEager` usage
- [x] Investigate "preview"
- [x] CSS entries exist in the manifest

## Related PRs
- https://github.com/laravel/laravel/pull/5944
- https://github.com/laravel/breeze/pull/171
- https://github.com/laravel/jetstream/pull/1089
- https://github.com/laravel/docs/pull/8049

## Things of note for upgrading projects

- SSR output build has changed extension: `bootstrap/ssr/ssr.mjs`. Although we could force this to remain as `.js`, [it looks like all future enhancements will be around modules](https://vitejs.dev/guide/ssr.html#ssr-format), so I think we should stick with this as the default IMO.
- `globEager` has been deprecated. You should use `import.meta.glob('pattern/here/**', { eager: true })`.
- Vite no longer creates self-signed certificates and need to be manually specified (note: because of this I think there is a reason to now look at adding support for detecting Valet certs as a follow up feature).

See more details on the Vite migration guide: https://vitejs.dev/guide/migration.html

### CODE REVIEW

**Comment by timacdonald** [comment]
Updated console output....

Screen Shot 2022-07-14 at 12 15 56 pm

File: src/index.ts [hunk 1]
```diff
@@ -159,9 +159,9 @@ function resolveLaravelPlugin(pluginConfig: Required<PluginConfig>): LaravelPlug
                     fs.writeFileSync(hotFile, viteDevServerUrl)
 
                     setTimeout(() => {
-                        server.config.logger.info(colors.red(`\n  Laravel ${laravelVersion()} `))
-                        server.config.logger.info(`\n  > APP_URL: ` + colors.cyan(appUrl))
-                    })
+                        server.config.logger.info(colors.red(`\n  ${colors.bold('Laravel')} ${laravelVersion()} `))
+                        server.config.logger.info(`  ${colors.green('➜')}  ${colors.bold('APP_URL')}: ${colors.cyan(appUrl.replace(/:(\d+)/, (_, port) => `:${colors.bold(port)}`))}`)
+                    }, 100)
```
Comment (timacdonald) [inline src/index.ts:164]: I hate that I'm specifying a timeout here, however I was unable to get it to work without something here and was unable to find any additional hooks

File: src/index.ts [hunk 2]
```diff
@@ -413,7 +348,10 @@ function noExternalInertiaHelpers(config: UserConfig): true|Array<string|RegExp>
     /* eslint-disable-next-line @typescript-eslint/ban-ts-comment */
     /* @ts-ignore */
     const userNoExternal = (config.ssr as SSROptions|undefined)?.noExternal
-    const pluginNoExternal = ['laravel-vite-plugin']
+    const pluginNoExternal = [
+        'laravel-vite-plugin/inertia-helpers',
+        '@inertiajs/server'
```
Comment (timacdonald) [inline src/index.ts:353]: This is likely something we need to push upstream and fix in inertia, but putting it here so I can move on.

File: src/index.ts [hunk 3]
```diff
@@ -183,7 +182,7 @@ function resolveLaravelPlugin(pluginConfig: Required<PluginConfig>): LaravelPlug
             exitHandlersBound = true
 
             return () => server.middlewares.use((req, res, next) => {
-                if (req.url === '/index.html') {
+                if (req.url === '/index.html' && req.headers.accept?.includes('html')) {
```
Comment (timacdonald) [inline src/index.ts:185]: sneaking this one in 🤫

File: src/index.ts [hunk 4]
```diff
@@ -159,9 +158,9 @@ function resolveLaravelPlugin(pluginConfig: Required<PluginConfig>): LaravelPlug
                     fs.writeFileSync(hotFile, viteDevServerUrl)
 
                     setTimeout(() => {
-                        server.config.logger.info(colors.red(`\n  Laravel ${laravelVersion()} `))
-                        server.config.logger.info(`\n  > APP_URL: ` + colors.cyan(appUrl))
-                    })
+                        server.config.logger.info(colors.red(`\n  ${colors.bold('Laravel')} ${laravelVersion()} `))
```
[the remark this answers is not in the document]
Comment (jessarcher) [reply src/index.ts]: Done :) I've also added the plugin version as I think that will help a lot when people send screenshots!

**Comment by timacdonald** [comment]
Thanks for the review @innocenzi ❤️

**Pull request merged**